runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/PortablePipelineJarCreator.java

```diff
@@ -0,0 +1,289 @@
+  @VisibleForTesting
+  protected static void copyResourcesFromJar(JarFile inputJar, JarOutputStream outputStream)
+      throws IOException {
+    Enumeration<JarEntry> inputJarEntries = inputJar.entries();
+    // The zip spec allows multiple files with the same name; the Java zip libraries do not.
+    // Keep track of the files we've already written to filter out duplicates.
+    // Also, ignore the old manifest; we want to write our own.
+    Set<String> previousEntryNames = new HashSet<>(ImmutableList.of(JarFile.MANIFEST_NAME));
+    while (inputJarEntries.hasMoreElements()) {
+      JarEntry inputJarEntry = inputJarEntries.nextElement();
+      InputStream inputStream = inputJar.getInputStream(inputJarEntry);
+      String entryName = inputJarEntry.getName();
+      if (previousEntryNames.contains(entryName)) {
+        LOG.debug("Skipping duplicated file {}", entryName);
+      } else {
+        JarEntry outputJarEntry = new JarEntry(inputJarEntry);
+        outputStream.putNextEntry(outputJarEntry);
+        LOG.trace("Copying jar entry {}", inputJarEntry);
+        IOUtils.copy(inputStream, outputStream);
+        previousEntryNames.add(entryName);
+      }
+    }
+  }
+
+  private static void writePipeline(Pipeline pipeline, JarOutputStream outputStream)
+      throws IOException {
+    JarEntry jarEntry = new JarEntry(PortablePipelineJarUtils.PIPELINE_FILE_NAME);
+    outputStream.putNextEntry(jarEntry);
+    pipeline.writeTo(outputStream);
+  }
+
+  private static void writePipelineOptions(Struct optionsProto, JarOutputStream outputStream)
+      throws IOException {
+    JarEntry jarEntry = new JarEntry(PortablePipelineJarUtils.PIPELINE_OPTIONS_FILE_NAME);
+    outputStream.putNextEntry(jarEntry);
+    optionsProto.writeTo(outputStream);
+  }
+
+  @VisibleForTesting
+  interface ArtifactRetriever {
+    GetManifestResponse getManifest(GetManifestRequest request);
+
+    Iterator<ArtifactChunk> getArtifact(GetArtifactRequest request);
+  }
+
+  /**
+   * Copy all artifacts retrievable via the {@link ArtifactRetrievalServiceBlockingStub} to the
+   * {@code outputStream}.
+   *
+   * @return A {@link ProxyManifest} pointing to the artifacts' location in the output jar.
+   */
+  @VisibleForTesting
+  static ProxyManifest copyStagedArtifacts(
+      String retrievalToken, JarOutputStream outputStream, ArtifactRetriever retrievalServiceStub)
+      throws IOException {
+    GetManifestRequest manifestRequest =
+        GetManifestRequest.newBuilder().setRetrievalToken(retrievalToken).build();
+    ArtifactApi.Manifest manifest = retrievalServiceStub.getManifest(manifestRequest).getManifest();
+    // Create a new proxy manifest to locate artifacts at jar runtime.
+    ProxyManifest.Builder proxyManifestBuilder = ProxyManifest.newBuilder().setManifest(manifest);
+    for (ArtifactMetadata artifact : manifest.getArtifactList()) {
+      String outputPath = PortablePipelineJarUtils.ARTIFACT_FOLDER_NAME + "/" + artifact.getName();
```

Comment:
The actual name of the artifact might have special characters so we should not put it in the path.
I think such issues can be avoided if we directly use JarArticactSageingService.

Reply:
In practice, arbitrary artifact names would already break the artifact system we have. For instance, the Python SDK [expects](https://github.com/apache/beam/blob/7ea4547a6f2530dffc06cb4dab3cdde58c2fd0dc/sdks/python/apache_beam/runners/worker/sdk_worker_main.py#L211-L212) a file named exactly `pickled_main_session`. This is provided by the Go bootstrap code, which writes [artifacts to files based on their name](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/artifact/materialize.go#L113).

Reply:
The artifact retrieval service will do the name translation for us. 
So in practice, artifacts are stored with arbitrary name (art001) but at the time of retrieval on worker machine, they are converted back to their original name.

Reply:
I guess I should have been more specific about what I meant by arbitrary artifact names.

While places like https://issues.apache.org/jira/browse/BEAM-4109 seem to indicate we want to support *artifact names which are not valid file names*, places like [here](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/artifact/materialize.go#L113) indicate that artifact names actually do have to be valid file names. Also, I don't see the advantage of having artifact names be distinct from file names, and it just makes things more confusing.

In light of that, I want to keep the artifacts saved under their actual names wherever possible instead of gibberish UUIDs, unless there's a compelling reason not to.

Reply:
I changed my mind and decided to use UUIDs after all.